feat(forge): add `vm.stopRecord`

tushar994 opened the pull request:
## Motivation
solves #10240 

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

DaniPopes reviewed (commented):
Thanks, this looks good! However, I realized that this might not cover all use cases: calling `stopRecordAndReturnAccesses` resets the accesses and it returns them for one address; if I want to stop and return for multiple addresses, this is not possible as `accesses` calls after will always return empty arrays;

I think we need a `stopRecord` that stops recording without clearing, and then clear on the next `record` call instead.

@grandizzy @mattsse thoughts?

tushar994 commented:
@grandizzy sorry for noob question, but how do I run the integration tests in `testdata/default/cheats` locally?

DaniPopes commented:
they are defined here https://github.com/foundry-rs/foundry/blob/master/crates/forge/tests/it/cheats.rs
`cargo test -p forge --test it test_cheats_local_default`

grandizzy commented:
> Thanks, this looks good! However, I realized that this might not cover all use cases: calling `stopRecordAndReturnAccesses` resets the accesses and it returns them for one address; if I want to stop and return for multiple addresses, this is not possible as `accesses` calls after will always return empty arrays;
> 
> I think we need a `stopRecord` that stops recording without clearing, and then clear on the next `record` call instead.
> 
> @grandizzy @mattsse thoughts?



> Thanks, this looks good! However, I realized that this might not cover all use cases: calling `stopRecordAndReturnAccesses` resets the accesses and it returns them for one address; if I want to stop and return for multiple addresses, this is not possible as `accesses` calls after will always return empty arrays;
> 
> I think we need a `stopRecord` that stops recording without clearing, and then clear on the next `record` call instead.
> 
> @grandizzy @mattsse thoughts?

Yeah, that makes sense. So `stopRecord` should not set `state.accesses` to None but rather set a flag to mark it as stopped/no new records allowed - that is to still be able to retrieve records after (would here make sense to have a `resumeRecord` that will reallow records in case one wants to skip some actions?). Then a new `record` call should set the flag to true and reset all recorded accesses.

tushar994 commented:
> > I think we need a `stopRecord` that stops recording without clearing, and then clear on the next `record` call instead.
> > @grandizzy @mattsse thoughts?
> 
> > Thanks, this looks good! However, I realized that this might not cover all use cases: calling `stopRecordAndReturnAccesses` resets the accesses and it returns them for one address; if I want to stop and return for multiple addresses, this is not possible as `accesses` calls after will always return empty arrays;
> > I think we need a `stopRecord` that stops recording without clearing, and then clear on the next `record` call instead.
> > @grandizzy @mattsse thoughts?
> 
> Yeah, that makes sense. So `stopRecord` should not set `state.accesses` to None but rather set a flag to mark it as stopped/no new records allowed - that is to still be able to retrieve records after (would here make sense to have a `resumeRecord` that will reallow records in case one wants to skip some actions?). Then a new `record` call should set the flag to true and reset all recorded accesses. @DaniPopes @tushar994 wdyt?

tushar994 commented:
> > I think we need a `stopRecord` that stops recording without clearing, and then clear on the next `record` call instead.
> > @grandizzy @mattsse thoughts?
> 
> > Thanks, this looks good! However, I realized that this might not cover all use cases: calling `stopRecordAndReturnAccesses` resets the accesses and it returns them for one address; if I want to stop and return for multiple addresses, this is not possible as `accesses` calls after will always return empty arrays;
> > I think we need a `stopRecord` that stops recording without clearing, and then clear on the next `record` call instead.
> > @grandizzy @mattsse thoughts?
> 
> Yeah, that makes sense. So `stopRecord` should not set `state.accesses` to None but rather set a flag to mark it as stopped/no new records allowed - that is to still be able to retrieve records after (would here make sense to have a `resumeRecord` that will reallow records in case one wants to skip some actions?). Then a new `record` call should set the flag to true and reset all recorded accesses. @DaniPopes @tushar994 wdyt?

I had thought the same, but didnt implement it because I thought i'd just stick to what was proposed in the issue. Will implement that! thank you!

tushar994 commented:
sorry closed it by accident

tushar994 commented:
Have added two cheatcodes
- resetRecord. This clears all records calls by accesses.
- stopRecord. This pauses recording of calls.
@grandizzy 

please review. thank you!!

DaniPopes reviewed (dismissed):
Thanks! I fixed some nits. I also removed `resetRecord` since we reset on the next `record` call anyway, and so `resetRecord` can be "implemented" using `vm.record(); vm.stopRecord();` if really necessary.

tushar994 commented:
> Thanks! I fixed some nits. I also removed `resetRecord` since we reset on the next `record` call anyway, and so `resetRecord` can be "implemented" using `vm.record(); vm.stopRecord();` if really necessary.

got it! makes sense. I thought keeping different commands for each keeps it more general, so it will accomodate more use cases. with the current implementation, theres no way to `pause` recording and then `unpause` with previous state intact. however, that also might not be a use case that anyone wants. thank you!!

tushar994 commented on testdata/default/cheats/Record.t.sol:
```diff
@@ -88,7 +88,7 @@ contract RecordTest is DSTest {
         vm.stopRecord();
```
can remove this right? doesnt serve any purpose is this test right?

grandizzy replied:
oops, missed that. it's harmless anyways so we can keep
